// @alchemy/aa-core/dist/esm/chains/index.js
import { defineChain } from "viem";
import { arbitrum as vab, arbitrumGoerli as vabg, arbitrumSepolia as vabs, base as vbase, baseGoerli as vbaseg, baseSepolia as vbases, goerli as vgo, mainnet as vmain, optimism as vop, optimismGoerli as vopg, optimismSepolia as vops, polygon as vpg, polygonMumbai as vpgm, polygonAmoy as vpga, sepolia as vsep, fraxtal as vfrax, zora as vzora, zoraSepolia as vzoras, } from "viem/chains";
export const arbitrum = {
    ...vab,
    rpcUrls: {
        ...vab.rpcUrls,
        alchemy: {
            http: ["https://arb-mainnet.g.alchemy.com/v2"],
        },
    },
};
export const arbitrumGoerli = {
    ...vabg,
    rpcUrls: {
        ...vabg.rpcUrls,
        alchemy: {
            http: ["https://arb-goerli.g.alchemy.com/v2"],
        },
    },
};
export const arbitrumSepolia = {
    ...vabs,
    rpcUrls: {
        ...vabs.rpcUrls,
        alchemy: {
            http: ["https://arb-sepolia.g.alchemy.com/v2"],
        },
    },
};
export const goerli = {
    ...vgo,
    rpcUrls: {
        ...vgo.rpcUrls,
        alchemy: {
            http: ["https://eth-goerli.g.alchemy.com/v2"],
        },
    },
};
export const mainnet = {
    ...vmain,
    rpcUrls: {
        ...vmain.rpcUrls,
        alchemy: {
            http: ["https://eth-mainnet.g.alchemy.com/v2"],
        },
    },
};
export const optimism = {
    ...vop,
    rpcUrls: {
        ...vop.rpcUrls,
        alchemy: {
            http: ["https://opt-mainnet.g.alchemy.com/v2"],
        },
    },
};
export const optimismGoerli = {
    ...vopg,
    rpcUrls: {
        ...vopg.rpcUrls,
        alchemy: {
            http: ["https://opt-goerli.g.alchemy.com/v2"],
        },
    },
};
export const optimismSepolia = {
    ...vops,
    rpcUrls: {
        ...vops.rpcUrls,
        alchemy: {
            http: ["https://opt-sepolia.g.alchemy.com/v2"],
        },
    },
};
export const sepolia = {
    ...vsep,
    rpcUrls: {
        ...vsep.rpcUrls,
        alchemy: {
            http: ["https://eth-sepolia.g.alchemy.com/v2"],
        },
    },
};
export const base = {
    ...vbase,
    rpcUrls: {
        ...vbase.rpcUrls,
        alchemy: {
            http: ["https://base-mainnet.g.alchemy.com/v2"],
        },
    },
};
export const baseGoerli = {
    ...vbaseg,
    rpcUrls: {
        ...vbaseg.rpcUrls,
        alchemy: {
            http: ["https://base-goerli.g.alchemy.com/v2"],
        },
    },
};
export const baseSepolia = {
    ...vbases,
    rpcUrls: {
        ...vbases.rpcUrls,
        alchemy: {
            http: ["https://base-sepolia.g.alchemy.com/v2"],
        },
    },
};
export const polygonMumbai = {
    ...vpgm,
    rpcUrls: {
        ...vpgm.rpcUrls,
        alchemy: {
            http: ["https://polygon-mumbai.g.alchemy.com/v2"],
        },
    },
};
export const polygonAmoy = {
    ...vpga,
    rpcUrls: {
        ...vpga.rpcUrls,
        alchemy: {
            http: ["https://polygon-amoy.g.alchemy.com/v2"],
        },
    },
};
export const polygon = {
    ...vpg,
    rpcUrls: {
        ...vpg.rpcUrls,
        alchemy: {
            http: ["https://polygon-mainnet.g.alchemy.com/v2"],
        },
    },
};
export const fraxtal = {
    ...vfrax,
    rpcUrls: {
        ...vfrax.rpcUrls,
    },
};
export const fraxtalSepolia = defineChain({
    id: 2523,
    name: "Fraxtal Sepolia",
    nativeCurrency: { name: "Frax Ether", symbol: "frxETH", decimals: 18 },
    rpcUrls: {
        default: {
            http: ["https://rpc.testnet-sepolia.frax.com"],
        },
    },
});
export const zora = {
    ...vzora,
    rpcUrls: {
        ...vzora.rpcUrls,
    },
};
export const zoraSepolia = {
    ...vzoras,
    rpcUrls: {
        ...vzoras.rpcUrls,
    },
};
