// @alchemy/aa-core/dist/cjs/client/decorators/bundlerClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.bundlerActions = void 0;
const estimateUserOperationGas_js_1 = require("../../actions/bundler/estimateUserOperationGas.js");
const getSupportedEntryPoints_js_1 = require("../../actions/bundler/getSupportedEntryPoints.js");
const getUserOperationByHash_js_1 = require("../../actions/bundler/getUserOperationByHash.js");
const getUserOperationReceipt_js_1 = require("../../actions/bundler/getUserOperationReceipt.js");
const sendRawUserOperation_js_1 = require("../../actions/bundler/sendRawUserOperation.js");
const bundlerActions = (client) => ({
    estimateUserOperationGas: async (request, entryPoint, stateOverride) => (0, estimateUserOperationGas_js_1.estimateUserOperationGas)(client, { request, entryPoint, stateOverride }),
    sendRawUserOperation: async (request, entryPoint) => (0, sendRawUserOperation_js_1.sendRawUserOperation)(client, { request, entryPoint }),
    getUserOperationByHash: async (hash) => (0, getUserOperationByHash_js_1.getUserOperationByHash)(client, { hash }),
    getSupportedEntryPoints: async () => (0, getSupportedEntryPoints_js_1.getSupportedEntryPoints)(client),
    getUserOperationReceipt: async (hash) => (0, getUserOperationReceipt_js_1.getUserOperationReceipt)(client, { hash }),
});
exports.bundlerActions = bundlerActions;
