// @alchemy/aa-core/dist/cjs/client/decorators/smartAccountClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.smartAccountClientMethodKeys = exports.smartAccountClientActions = void 0;
const buildUserOperation_js_1 = require("../../actions/smartAccount/buildUserOperation.js");
const buildUserOperationFromTx_js_1 = require("../../actions/smartAccount/buildUserOperationFromTx.js");
const buildUserOperationFromTxs_js_1 = require("../../actions/smartAccount/buildUserOperationFromTxs.js");
const checkGasSponsorshipEligibility_js_1 = require("../../actions/smartAccount/checkGasSponsorshipEligibility.js");
const dropAndReplaceUserOperation_js_1 = require("../../actions/smartAccount/dropAndReplaceUserOperation.js");
const getAddress_js_1 = require("../../actions/smartAccount/getAddress.js");
const sendTransaction_js_1 = require("../../actions/smartAccount/sendTransaction.js");
const sendTransactions_js_1 = require("../../actions/smartAccount/sendTransactions.js");
const sendUserOperation_js_1 = require("../../actions/smartAccount/sendUserOperation.js");
const signMessage_js_1 = require("../../actions/smartAccount/signMessage.js");
const signMessageWith6492_js_1 = require("../../actions/smartAccount/signMessageWith6492.js");
const signTypedData_js_1 = require("../../actions/smartAccount/signTypedData.js");
const signTypedDataWith6492_js_1 = require("../../actions/smartAccount/signTypedDataWith6492.js");
const signUserOperation_js_1 = require("../../actions/smartAccount/signUserOperation.js");
const upgradeAccount_js_1 = require("../../actions/smartAccount/upgradeAccount.js");
const waitForUserOperationTransacation_js_1 = require("../../actions/smartAccount/waitForUserOperationTransacation.js");
const smartAccountClientActions = (client) => ({
    buildUserOperation: (args) => (0, buildUserOperation_js_1.buildUserOperation)(client, args),
    buildUserOperationFromTx: (args, overrides, context) => (0, buildUserOperationFromTx_js_1.buildUserOperationFromTx)(client, args, overrides, context),
    buildUserOperationFromTxs: (args) => (0, buildUserOperationFromTxs_js_1.buildUserOperationFromTxs)(client, args),
    checkGasSponsorshipEligibility: (args) => (0, checkGasSponsorshipEligibility_js_1.checkGasSponsorshipEligibility)(client, args),
    signUserOperation: (args) => (0, signUserOperation_js_1.signUserOperation)(client, args),
    dropAndReplaceUserOperation: (args) => (0, dropAndReplaceUserOperation_js_1.dropAndReplaceUserOperation)(client, args),
    sendTransaction: (args, overrides, context) => (0, sendTransaction_js_1.sendTransaction)(client, args, overrides, context),
    sendTransactions: (args) => (0, sendTransactions_js_1.sendTransactions)(client, args),
    sendUserOperation: (args) => (0, sendUserOperation_js_1.sendUserOperation)(client, args),
    waitForUserOperationTransaction: (args) => waitForUserOperationTransacation_js_1.waitForUserOperationTransaction.bind(client)(client, args),
    upgradeAccount: (args) => (0, upgradeAccount_js_1.upgradeAccount)(client, args),
    getAddress: (args) => (0, getAddress_js_1.getAddress)(client, args),
    signMessage: (args) => (0, signMessage_js_1.signMessage)(client, args),
    signTypedData: (args) => (0, signTypedData_js_1.signTypedData)(client, args),
    signMessageWith6492: (args) => (0, signMessageWith6492_js_1.signMessageWith6492)(client, args),
    signTypedDataWith6492: (args) => (0, signTypedDataWith6492_js_1.signTypedDataWith6492)(client, args),
});
exports.smartAccountClientActions = smartAccountClientActions;
exports.smartAccountClientMethodKeys = Object.keys((0, exports.smartAccountClientActions)(undefined)).reduce((accum, curr) => {
    accum.add(curr);
    return accum;
}, new Set());
