// @alchemy/aa-core/dist/cjs/chains/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.zoraSepolia = exports.zora = exports.fraxtalSepolia = exports.fraxtal = exports.polygon = exports.polygonAmoy = exports.polygonMumbai = exports.baseSepolia = exports.baseGoerli = exports.base = exports.sepolia = exports.optimismSepolia = exports.optimismGoerli = exports.optimism = exports.mainnet = exports.goerli = exports.arbitrumSepolia = exports.arbitrumGoerli = exports.arbitrum = void 0;
const viem_1 = require("viem");
const chains_1 = require("viem/chains");
exports.arbitrum = {
    ...chains_1.arbitrum,
    rpcUrls: {
        ...chains_1.arbitrum.rpcUrls,
        alchemy: {
            http: ["https://arb-mainnet.g.alchemy.com/v2"],
        },
    },
};
exports.arbitrumGoerli = {
    ...chains_1.arbitrumGoerli,
    rpcUrls: {
        ...chains_1.arbitrumGoerli.rpcUrls,
        alchemy: {
            http: ["https://arb-goerli.g.alchemy.com/v2"],
        },
    },
};
exports.arbitrumSepolia = {
    ...chains_1.arbitrumSepolia,
    rpcUrls: {
        ...chains_1.arbitrumSepolia.rpcUrls,
        alchemy: {
            http: ["https://arb-sepolia.g.alchemy.com/v2"],
        },
    },
};
exports.goerli = {
    ...chains_1.goerli,
    rpcUrls: {
        ...chains_1.goerli.rpcUrls,
        alchemy: {
            http: ["https://eth-goerli.g.alchemy.com/v2"],
        },
    },
};
exports.mainnet = {
    ...chains_1.mainnet,
    rpcUrls: {
        ...chains_1.mainnet.rpcUrls,
        alchemy: {
            http: ["https://eth-mainnet.g.alchemy.com/v2"],
        },
    },
};
exports.optimism = {
    ...chains_1.optimism,
    rpcUrls: {
        ...chains_1.optimism.rpcUrls,
        alchemy: {
            http: ["https://opt-mainnet.g.alchemy.com/v2"],
        },
    },
};
exports.optimismGoerli = {
    ...chains_1.optimismGoerli,
    rpcUrls: {
        ...chains_1.optimismGoerli.rpcUrls,
        alchemy: {
            http: ["https://opt-goerli.g.alchemy.com/v2"],
        },
    },
};
exports.optimismSepolia = {
    ...chains_1.optimismSepolia,
    rpcUrls: {
        ...chains_1.optimismSepolia.rpcUrls,
        alchemy: {
            http: ["https://opt-sepolia.g.alchemy.com/v2"],
        },
    },
};
exports.sepolia = {
    ...chains_1.sepolia,
    rpcUrls: {
        ...chains_1.sepolia.rpcUrls,
        alchemy: {
            http: ["https://eth-sepolia.g.alchemy.com/v2"],
        },
    },
};
exports.base = {
    ...chains_1.base,
    rpcUrls: {
        ...chains_1.base.rpcUrls,
        alchemy: {
            http: ["https://base-mainnet.g.alchemy.com/v2"],
        },
    },
};
exports.baseGoerli = {
    ...chains_1.baseGoerli,
    rpcUrls: {
        ...chains_1.baseGoerli.rpcUrls,
        alchemy: {
            http: ["https://base-goerli.g.alchemy.com/v2"],
        },
    },
};
exports.baseSepolia = {
    ...chains_1.baseSepolia,
    rpcUrls: {
        ...chains_1.baseSepolia.rpcUrls,
        alchemy: {
            http: ["https://base-sepolia.g.alchemy.com/v2"],
        },
    },
};
exports.polygonMumbai = {
    ...chains_1.polygonMumbai,
    rpcUrls: {
        ...chains_1.polygonMumbai.rpcUrls,
        alchemy: {
            http: ["https://polygon-mumbai.g.alchemy.com/v2"],
        },
    },
};
exports.polygonAmoy = {
    ...chains_1.polygonAmoy,
    rpcUrls: {
        ...chains_1.polygonAmoy.rpcUrls,
        alchemy: {
            http: ["https://polygon-amoy.g.alchemy.com/v2"],
        },
    },
};
exports.polygon = {
    ...chains_1.polygon,
    rpcUrls: {
        ...chains_1.polygon.rpcUrls,
        alchemy: {
            http: ["https://polygon-mainnet.g.alchemy.com/v2"],
        },
    },
};
exports.fraxtal = {
    ...chains_1.fraxtal,
    rpcUrls: {
        ...chains_1.fraxtal.rpcUrls,
    },
};
exports.fraxtalSepolia = (0, viem_1.defineChain)({
    id: 2523,
    name: "Fraxtal Sepolia",
    nativeCurrency: { name: "Frax Ether", symbol: "frxETH", decimals: 18 },
    rpcUrls: {
        default: {
            http: ["https://rpc.testnet-sepolia.frax.com"],
        },
    },
});
exports.zora = {
    ...chains_1.zora,
    rpcUrls: {
        ...chains_1.zora.rpcUrls,
    },
};
exports.zoraSepolia = {
    ...chains_1.zoraSepolia,
    rpcUrls: {
        ...chains_1.zoraSepolia.rpcUrls,
    },
};
